platform/darwin/src/MGLMapSnapshotter.mm

```diff
@@ -505,59 +449,70 @@ + (MGLImage*)drawAttributedSnapshotWorker:(mbgl::MapSnapshotter::Attributions)at
 #endif
 }
 
-- (void)drawAttributedSnapshot:(mbgl::MapSnapshotter::Attributions)attributions snapshotImage:(MGLImage *)mglImage pointForFn:(mbgl::MapSnapshotter::PointForFn)pointForFn latLngForFn:(mbgl::MapSnapshotter::LatLngForFn)latLngForFn overlayHandler:(MGLMapSnapshotOverlayHandler)overlayHandler {
+- (void)decorateImage:(MGLImage *)mglImage withOptions:(MGLMapSnapshotOptions *)options attributions:(mbgl::MapSnapshotter::Attributions)attributions pointForFn:(mbgl::MapSnapshotter::PointForFn)pointForFn latLngForFn:(mbgl::MapSnapshotter::LatLngForFn)latLngForFn overlayHandler:(MGLMapSnapshotOverlayHandler)overlayHandler completionHandler:(MGLMapSnapshotCompletionHandler)completion queue:(dispatch_queue_t)resultQueue {
 
     // Process image watermark in a work queue
     dispatch_queue_t workQueue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
-    dispatch_queue_t resultQueue = self.resultQueue;
-
-    // Capture scale and size by value to avoid accessing self from another thread
-    CGFloat scale  = self.options.scale;
-    CGSize size    = self.options.size;
-    BOOL showsLogo = self.options.showsLogo;
-
-    // pointForFn is a copyable std::function that captures state by value: see MapSnapshotter::Impl::snapshot
-    __weak __typeof__(self) weakself = self;
 
+    __weak __typeof__(self) weakSelf = self;
     dispatch_async(workQueue, ^{
```

Comment:
Do we really need to decorate the image on a worker thread? `MGLAttributedSnapshot()` doesn’t even show up in time profiles of a basic snapshot at 1512 × 1288 points. Removing this worker thread dance would greatly simplify the code: we could make this method synchronous and have it call a completion block that dispatches to the result queue.

Reply:
Hi there. Just looking at https://github.com/mapbox/mapbox-gl-native-ios/blob/bdb2e61624b30b57fcbb9b9c38540526aeab364f/platform/darwin/src/MGLMapSnapshotter.mm#L274

It looks like this decoration is happening on the same queue that the snapshot is occurring on? Is that correct? If that's the case I don't see much of a reason unless that queue is expected to be busy.

Reply:
Agreed - the only thing I wonder is if the customOverlay rendering takes time...

Reply:
Indeed, that comment is outdated. The block passed into `MapSnapshotter::snapshot()` is executed on the same queue that `-[MGLMapSnapshotter startWithQueue:completionHandler:]` is called from, though the actual resource fetching and rendering happens on a background thread as usual. `-decorateImage:withOptions:attributions:pointForFn:latLngForFn:overlayHandler:completionHandler:queue:` does the decoration on a different background queue, including running `overlayHandler`, and finally the results are dispatched to the queue passed into `-startWithQueue:completionHandler:`. In all, there are three queues involved just at the SDK level, not counting what happens internally to mbgl.

The `overlayHandler` could take arbitrary time depending on what the developer specifies. If we remove `workQueue` from the picture, would it make sense to run the overlayHandler on the initial queue that calls `-startWithQueue:completionHandler:` or on the queue passed into that method?

Reply:
> MGLAttributedSnapshot() doesn’t even show up in time profiles of a basic snapshot at 1512 × 1288 points.

I don't have context on what this is, but we're about to greatly expand what can be done to snapshots with runtime styling. It seems like this is the wrong time to move everything to the main thread. 

> Removing this worker thread dance would greatly simplify the code: we could make this method synchronous and have it call a completion block that dispatches to the result queue.

Then snapshot generation could potentially block the main thread, right? Not sure this is a good idea. There are use cases where 6-7 of these snapshots are being generated at the same time.

Reply:
Oh, sorry if I responded to an outdated comment.

> would it make sense to run the overlayHandler on the initial queue that calls -startWithQueue:completionHandler: or on the queue passed into that method?

I think this makes sense.

Reply:
As of mapbox/mapbox-gl-native#16268, the block that’s literally defined inline inside of `-[MGLMapSnapshotter startWithQueue:completionHandler:]` does run on the queue on which `-startWithQueue:completionHandler:` is called, typically the main queue. That change took place so that the developer can safely implement the runtime styling hooks called for in #200 without accidentally creating race conditions between the main thread and some background thread they have no awareness of. However, the bulk of the work continues to occur on background threads.

What I’m still unsure of is which queue the developer expects the overlay handler to be called on:

* The initial queue on which `-startWithQueue:completionHandler:` is called
* The special-purpose worker queue that the developer has no control over (current behavior)
* The queue that the developer passed into `-startWithQueue:completionHandler:` ostensibly for the purpose of receiving the finished snapshot

Despite the amount of code involved, `MGLAttributedSnapshot()` and the methods it calls are so trivial that they can comfortably occur on any thread. But the `overlayHandler` callback is less certain because the developer has free rein to perform expensive tasks in that callback.

Reply:
For now, I’ve kept the decorator queue but consolidated the dispatching into a single method, making the other methods synchronous on whatever queue they’re called on. This will simplify any potential change that would keep the snapshotter alive until completion.

Reply:
Just my 2 cents: 

I think it's fine to utilize the queue the developer passed in for the overlay handler work with a few caveats:
1) They could pass in the main queue. This, in my very opinionated view, would be a  developer error in this context and I would assert on it. However given this is an existing API, it's hard to assert on it without probably generating support tickets 🙃. Plus many developers (myself included) use the main queue as a sync point for completion handlers.  
2) Related to 1 - I'm unfamiliar with the queues mbgl is using but the interplay of multiple queues here has me looking at potential priority inversion issues. If the developer passes in a queue at the utility QOS, it'll slow the whole process down. Just something to keep in mind.

Tangentially related to this:
https://github.com/mapbox/mapbox-gl-native-ios/blob/master/platform/darwin/src/MGLMapSnapshotter.h#L241

That function should document what queue the completion handler is called on when developer's don't pass in a queue.